chan.go

```diff
@@ -145,6 +149,10 @@ func (re *chanResponseEmitter) Emit(v interface{}) error {
 
 	ctx := re.req.Context
 
+	if err := ctx.Err(); err != nil {
+		return err
+	}
+
```

Comment:
The declaration of ctx, and this check should probably be hoisted to the top of the function.
If we're canceled, we probably shouldn't do anything other than return.
The locking, EmitChan, etc. seems pointless if we'll get here anyway.

Reply:
I'd rather just leave it next to the `select` like this, so that we preserve the existing post-cancellation side effects of consuming v if it's a chan, and unblocking `Length()`, which might be important somewhere else (don't know off the top of my head).